File: java/90.subsets-ii.java
/*
 * @lc app=leetcode id=90 lang=java
 *
 * [90] Subsets II
 */

// @lc code=start
class Solution {

    List<List<Integer>> res = new LinkedList<>();

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        backtrack(0, nums, new ArrayList<>());
        return res;
    }

    private void backtrack(int s, int[] nums, List<Integer> tmpList) {
        res.add(new ArrayList<>(tmpList));
        for (int i = s; i < nums.length; ++i) {
            if (i > s && nums[i] == nums[i-1]) continue;
            tmpList.add(nums[i]);
            backtrack(i+1, nums, tmpList);
            tmpList.remove(tmpList.size()-1);
        }
    }
}
// @lc code=end

File: go/90.subsets-ii.go
/*
 * @lc app=leetcode id=90 lang=golang
 *
 * [90] Subsets II
 */

// @lc code=start
func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	res := [][]int{}
	var backtrack func(int, []int)
	backtrack = func(idx int, comb []int) {
		res = append(res, append([]int{}, comb...))

		for i := idx; i < len(nums); i++ {
			comb = append(comb, nums[i])
			backtrack(i+1, comb)
			comb = comb[:len(comb)-1]
			for i+1 < len(nums) && nums[i+1] == nums[i] {
				i++
			}
		}
	}

	backtrack(0, []int{})
	return res
}
// @lc code=end

